File: week4/c++/week4_15.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

int main(void)
{

    while (1)
    {
        string in;
        getline(cin, in);

        // 종료
        if (in.length() == 1 && in[0] == '.')
        {
            break;
        }

        // 판단
        stack<char> bracket;
        bool isRight = true;
        for (int i = 0; i < in.length(); i++)
        {
            if (in[i] == '(' || in[i] == '[')
            {
                bracket.push(in[i]);
            }
            else if (in[i] == ')')
            {
                if (!bracket.empty() && bracket.top() == '(')
                {
                    bracket.pop();
                }
                else
                {
                    isRight = false;
                    break;
                }
            }
            else if (in[i] == ']')
            {
                if (!bracket.empty() && bracket.top() == '[')
                {
                    bracket.pop();
                }
                else
                {
                    isRight = false;
                    break;
                }
            }
        }

        if (isRight && bracket.empty())
        {
            cout << "yes" << endl;
        }
        else
        {
            cout << "no" << endl;
        }
    }
    return 0;
}

File: week4/java/week4_15.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.Stack;

public class week4_15 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        ArrayList<String> list = new ArrayList<String>();
        while(true) {
            String str = br.readLine();
            if(str.charAt(0)=='.') break;
            list.add(str);
        }

        Iterator iter = list.iterator();

        while(iter.hasNext()) {
            String str = (String)iter.next();

            if(str.charAt(0)=='.') break;

            Stack<Character> stack = new Stack<Character>();

            boolean result = true;
            for(int i = 0; i < str.length() ; i++) {
                if(str.charAt(i)=='(') {
                    stack.push(str.charAt(i));
                }
                else if(str.charAt(i)==')') {
                    if(!stack.isEmpty()){
                        if(!(stack.peek()=='(')) break;
                        else stack.pop();
                    }else{
                        result = false;
                        break;
                    }
                }
                else if(str.charAt(i)=='[') {
                    stack.push(str.charAt(i));
                }
                else if(str.charAt(i)==']') {
                    if(!stack.isEmpty()){
                        if(!(stack.peek()=='[')) break;
                        else stack.pop();
                    }else{
                        result = false;
                        break;
                    }
                }
            }

            if(stack.isEmpty() && result)
                System.out.println("yes");
            else
                System.out.println("no");
        }
    }
}
